In [1]:
import numpy as np

In [4]:
# define a function that will read in a given file 
# From Assign 1
def readfile(filename):
    
    # open the file 
    f = open(filename,'r')
    
    # read header info line by line (line will be a string)
    # read first two lines FIRST 
    
    # read the time 
    line = f.readline()
    label, value = line.split()
    time = float(value)

    # read the number of particles 
    line = f.readline()
    label, value = line.split()
    total = float(value)
    
    # close file
    f.close()

    # read the remainder of the file, skipping the first 3 lines 
    # dtype: delimiter is white space 
    # names=True sets columns as arrays with names specified by 
    # line 4 of the file 
    data = np.genfromtxt(filename, dtype=None, names=True, skip_header=3)
    
    # this will return the time of the snapshot,
    # total number of particles 
    #and an array that stores the remainder of the data. 
    return time, total, data

In [2]:
# Question 1 
# function that computes the mass of a given component of 
# a given galaxy 
# returns the mass in units of 1e12
# input, filename to store the galaxy name
# input, Ptype
# input, data array already loaded in

def Mass(filename,Ptype,data):
    
    # identify the galaxy name
    splitting = filename.split("_")
    galname = splitting[0]
    
    # identify all particles of required type
    index = np.where(data['type']== Ptype)
    
    # compute the total mass of desired particles
    mass = np.sum(data['m'][index])
    # store mass in units of 1e12 
    mass1e12 = round(mass*1e10/1e12,3)
    
    # compute the total number of particles
    number = data['m'][index].size
     
    print "The Total Mass of particle type %s in the galaxy %s is" %(Ptype,galname), mass1e12, "x 10^12 Msun"
    print "The Total Number of particle type %s in the galaxy %s is" %(Ptype,galname), number

    return mass1e12

In [5]:
timeMW, totalMW, dataMW = readfile('MW_000.txt')

In [6]:
timeM31, totalM31, dataM31 = readfile('M31_000.txt')

In [7]:
timeM33, totalM33, dataM33 = readfile('M33_000.txt')

In [8]:
# Question 2 
# Print the mass of the Milky Way 
print "Halo"
HaloMW = Mass("MW_000.txt",1,dataMW)
print "\n"
print "Disk"
DiskMW = Mass("MW_000.txt",2,dataMW)
print "\n"
print "Bulge"
BulgeMW = Mass("MW_000.txt",3,dataMW)

Halo
The Total Mass of particle type 1 in the galaxy MW is 1.975 x 10^12 Msun
The Total Number of particle type 1 in the galaxy MW is 50000


Disk
The Total Mass of particle type 2 in the galaxy MW is 0.075 x 10^12 Msun
The Total Number of particle type 2 in the galaxy MW is 75000


Bulge
The Total Mass of particle type 3 in the galaxy MW is 0.01 x 10^12 Msun
The Total Number of particle type 3 in the galaxy MW is 10000


In [9]:
# Question 2 
# Print the mass of the M31
print "Halo"
HaloM31 = Mass("M31_000.txt",1,dataM31)
print "\n"
print "Disk"
DiskM31 = Mass("M31_000.txt",2,dataM31)
print "\n"
print "Bulge"
BulgeM31 = Mass("M31_000.txt",3,dataM31)

Halo
The Total Mass of particle type 1 in the galaxy M31 is 1.921 x 10^12 Msun
The Total Number of particle type 1 in the galaxy M31 is 50000


Disk
The Total Mass of particle type 2 in the galaxy M31 is 0.12 x 10^12 Msun
The Total Number of particle type 2 in the galaxy M31 is 120000


Bulge
The Total Mass of particle type 3 in the galaxy M31 is 0.019 x 10^12 Msun
The Total Number of particle type 3 in the galaxy M31 is 19000


In [10]:
# Question 2 
# Print the mass of the M33  No Bulge
print "Halo"
HaloM33 = Mass("M33_000.txt",1,dataM33)
print "\n"
print "Disk"
DiskM33 = Mass("M33_000.txt",2,dataM33)
print "\n"
print "Bulge"
TotM33 = HaloM33+DiskM33
print "total mass M33", TotM33, "x 10^12 Msun"

Halo
The Total Mass of particle type 1 in the galaxy M33 is 0.187 x 10^12 Msun
The Total Number of particle type 1 in the galaxy M33 is 5000


Disk
The Total Mass of particle type 2 in the galaxy M33 is 0.009 x 10^12 Msun
The Total Number of particle type 2 in the galaxy M33 is 9300


Bulge
total mass M33 0.196 x 10^12 Msun


In [22]:
# Question 3
# comparing the total mass of MW to M31
print "Ratio Between the Halo mass of the M31 and MW is", round((HaloMW)/(HaloM31),2)

Ratio Between the Halo mass of the M31 and MW is 1.03


They are roughly equal

In [21]:
# Question 4 
# total mass of the local Group
totMass = HaloM31 + HaloMW + HaloM33 + DiskM31 + DiskMW + DiskM33 + BulgeMW + BulgeM31
print "The total Mass of the Local Group", totMass, " x 10^12 Msun" 
# sort the masses of each component 
ListMass = [HaloM31,HaloMW,HaloM33,DiskM31,DiskMW,DiskM33,BulgeMW,BulgeM31]
print sorted(ListMass,reverse=True)

The total Mass of the Local Group 4.316  x 10^12 Msun
[1.975, 1.921, 0.187, 0.12, 0.075, 0.019, 0.01, 0.009]


Total Mass is dominated by the halo mass of M31 and the MW

In [13]:
# Question 5 
# comparing the stellar mass of MW to M31
ratioStar = round((DiskMW+BulgeMW)/(DiskM31+BulgeM31),2)
print "Ratio Between the stellar mass of M31 and MW ", ratioStar 
print "The Stellar mass of the MW is %s Percent that of M31" %(ratioStar*100)

Ratio Between the stellar mass of M31 and MW  0.61
The Stellar mass of the MW is 61.0 Percent that of M31


In [14]:
# Question 6 
# comparing the total dark mass of M33 and M31
print "Ratio between the halo mass of M31 and M33 is", round(HaloM31/HaloM33,2)
print "The Halo Mass of M31 is %s times larger" %round(HaloM31/HaloM33,2)

Ratio between the halo mass of M31 and M33 is 10.27
The Halo Mass of M31 is 10.27 times larger


In [15]:
# Question 7
# comparing the stellar mass of M33 to that of M31
ratioStarM33M31 = round((DiskM31+BulgeM31)/(DiskM33),2)
print "Ratio between the stellar mass of M31 and M33 is", ratioStarM33M31
print "The stellar ratio is %s times larger than that of the dark component" %(ratioStarM33M31)

Ratio between the stellar Mass of M31 and M33 is 15.44
The stellar ratio is 15.44 times larger than that of the dark component


In [16]:
# Question 8
# computing baryon fraction
RatioMW = round((DiskMW+BulgeMW)/HaloMW,2)
RatioM31 = round((DiskM31+BulgeM31)/HaloM31,2)
RatioM33 = round((DiskM33)/HaloM33,2)
print "The Ratio Light/Dark for the MW is", RatioMW*100, "percent"
print "The Ratio Light/Dark for the M31 is", RatioM31*100, "percent"
print "The Ratio Light/Dark for the M33 is", RatioM33*100, "percent"


The Ratio Light/Dark for the MW is 4.0 percent
The Ratio Light/Dark for the M31 is 7.0 percent
The Ratio Light/Dark for the M33 is 5.0 percent


These fractions are well below the universal baryon fraction of 15%.
Reasons: 1) Outflows may feed the CGM, removing baryons from the disk
2) Gas may be inefficiently accreted onto galactic disks